### PR TITLE
Seneca out-of-range temperature detection

### LINKED ISSUE #490: `SenecaK107`: Replace out-of-range temperatures with NaN
Currently, if one of the sensors is not attached to anything, its temperature will be given as >600°C. It would be better to catch these out-of-range values and replace them with NaNs so they don't skew the plots or data analysis.

Once UNIRAS is finished, it is unlikely that we will see this problem, but *if* a sensor wire were to become disconnected, it would be more useful to have the output be a NaN (possibly also with a warning in the program log).

### PR DESCRIPTION
**Pull request by CWestICL**
Added logic to `calc_temp()` that replaces temperatures outside the minimum and maximum temperature ranges with NaNs. If the resulting numpy array contains NaNs, a warning is raised.

Tests have also been altered to expect the NaN values when parsing the data.

Closes #490

### CODE REVIEW

File: tests/hardware/plugins/temperature/test_senecaK107.py [hunk 1]
```diff
@@ -87,12 +87,12 @@ def test_parse_data(dev: SenecaK107, data: bytes) -> None:
...
-        631.4406250000001,
-        631.4175,
+        float("nan"),
+        float("nan"),
         20.14281249999999,
-        631.4406250000001,
+        float("nan"),
         19.946250000000006,
-        631.4406250000001,
+        float("nan"),
```
Comment (jamesturner246) [inline tests/hardware/plugins/temperature/test_senecaK107.py:95]: Might also want to test to see if `MIN_TEMP` is respected here.

**Review by jamesturner246 — APPROVED** [review]
LGTM! See the note about `MIN_TEMP` below.

File: tests/hardware/plugins/temperature/test_senecaK107.py (hunk 1 above)
Comment (CWestICL) [reply tests/hardware/plugins/temperature/test_senecaK107.py:95]: Yeah this is definitely something we should write tests for, but I imagine it'll require having examples of bytestring data containing temperatures below the minimum limit. I don't know enough about bytestrings to fabricate test data like this so might be worth making a separate issue.

**Review by dc2917 — CHANGES_REQUESTED** [review]
It likely won't make any practical difference but I think we should use `numpy.nan` rather than `float("nan")`. As the values are all in a NumPy array anyway, I think it makes sense to use their NaN implementation.

File: finesse/hardware/plugins/temperature/senecak107.py [hunk 2]
```diff
@@ -146,6 +150,13 @@ def calc_temp(self, vals: numpy.ndarray) -> numpy.ndarray:
         calc *= self.SCALING_FACTOR
         # Adjusts for minimum temperature limit
         calc += self.MIN_TEMP
+
+        calc[calc > self.MAX_TEMP] = numpy.nan
+        calc[calc < self.MIN_TEMP] = numpy.nan
+
+        if numpy.isnan(calc).any():
+            logging.warn(f"Out-of-range temperature(s) detected: {calc}")
```
Comment (dc2917) [inline finesse/hardware/plugins/temperature/senecak107.py:158]: `logging.warn` seems to be undocumented so I'm assuming it's deprecated.
```suggestion
            logging.warning(f"Out-of-range temperature(s) detected: {calc}")
```

Also, I'm fairly sure calling `logging` to do the logging directly is advised against - we should be using `logging.getLogger()`. However, it seems everywhere else in the code `logging` is called directly. @dalonsoa do you have any insight on this? I know you used `logging.getLogger()` in your `materials` module in PyECN.

Comment (dalonsoa) [reply finesse/hardware/plugins/temperature/senecak107.py:158]: It is not advised again - [Python documentation uses it with no problem in the first example](https://docs.python.org/3/library/logging.html) - but might be not what you want.

So, `logging.warning`, `logging.error`, etc. all use the `root` logger. If we use `logging.getLogger()`, with no arguments, we also get the `root` logger, therefore `logging.warning("something")` is the same that  `logging.getLogger().warning("something")`.

The difference comes when you have different loggers configured for different parts of the program or want to configure a specific behavior depending on where the logging happen, or what the log is about. In that case you could do `logging.getLogger("gui").warning("something")`, for example, and have a handler configured that send all the logging events sent to the `gui` logger to a pop-up window or something like that.

In summary, in our case, it does not matter. In general, it might.

Comment (dc2917) [reply finesse/hardware/plugins/temperature/senecak107.py:158]: Ah okay, thanks for clarifying!

**Review by dc2917 — APPROVED** [review]
